File: src/main/java/me/rsman/BetterMinecraftCore/Managers/EnchantManager.java
package me.rsman.BetterMinecraftCore.Managers;

import me.rsman.BetterMinecraftCore.enums.EEnchants;
import org.bukkit.NamespacedKey;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemStack;

import java.lang.reflect.Field;
import java.util.Arrays;
import java.util.HashMap;
import java.util.stream.Collectors;

public final class EnchantManager {

    public static void registerEnchantment(Enchantment enchantment){
        try{
            Field f = Enchantment.class.getDeclaredField("acceptingNew");
            f.setAccessible(true);
            f.set(null, true);
            Enchantment.registerEnchantment(enchantment);
        } catch (Exception ignored){}
    }
    public static void registerAllEnchantments(){
        for (EEnchants enchantment : EEnchants.values()){
            if(!Arrays.stream(Enchantment.values()).collect(Collectors.toList()).contains(enchantment.getEnchant()))
                registerEnchantment(enchantment.getEnchant());
        }
    }
    public static void unRegisterAllEnchantments(){
        try {
            Field keyField = Enchantment.class.getDeclaredField("byKey");
            keyField.setAccessible(true);
            @SuppressWarnings("unchecked")
            HashMap<NamespacedKey, Enchantment> byKey = (HashMap<NamespacedKey, Enchantment>) keyField.get(null);

            Field nameField = Enchantment.class.getDeclaredField("byName");
            nameField.setAccessible(true);
            @SuppressWarnings("unchecked")
            HashMap<String, Enchantment> byName = (HashMap<String, Enchantment>) nameField.get(null);

            for (EEnchants enchantment : EEnchants.values()){
                if(enchantment.getKey().startsWith(NamespacedKey.minecraft("bmc_").toString())){
                    byKey.remove(enchantment.getEnchant().getKey());
                    byName.remove(enchantment.getEnchant().getName());
                }
            }
        } catch (Exception ignored) { }
    }
    public static boolean hasEnchantment(ItemStack item, Enchantment enchantment){
        return item.getEnchantments().containsKey(enchantment);
    }
    public static void addEnchantment(ItemStack item, Enchantment enchantment, int level){
        item.addUnsafeEnchantment(enchantment, level);
        ItemManager.updateItemLore(item);
    }
    public static void removeEnchantment(ItemStack item, Enchantment enchantment){
        item.removeEnchantment(enchantment);
        ItemManager.updateItemLore(item);
    }
    public static Integer getEnchantmentLevel(ItemStack item, Enchantment enchantment){
        return item.getEnchantments().get(enchantment);
    }
    public static boolean isCustom(Enchantment enchantment){
        return enchantment.getKey().toString().startsWith(NamespacedKey.minecraft("bmc_").toString());
    }
}

File: src/main/kotlin/me/rsman/BetterMinecraftCore/Managers/EnchantManager.kt
package me.rsman.BetterMinecraftCore.Managers

import java.util.HashMap
import org.bukkit.enchantments.Enchantment
import me.rsman.BetterMinecraftCore.enums.EEnchants
import me.rsman.BetterMinecraftCore.extensions.updateCustomLore
import org.bukkit.NamespacedKey
import java.util.Arrays
import org.bukkit.inventory.ItemStack
import java.util.stream.Collectors
import java.lang.Exception

object EnchantManager {
    fun registerEnchantment(enchantment: Enchantment?) {
        try {
            val f = Enchantment::class.java.getDeclaredField("acceptingNew")
            f.isAccessible = true
            f[null] = true
            Enchantment.registerEnchantment(enchantment!!)
        } catch (ignored: Exception) {
        }
    }

    fun registerAllEnchantments() {
        for (enchantment in EEnchants.values()) {
            if (!Arrays.stream(Enchantment.values()).collect(Collectors.toList()).contains(enchantment.enchant)) registerEnchantment(enchantment.enchant)
        }
    }

    fun unRegisterAllEnchantments() {
        try {
            val keyField = Enchantment::class.java.getDeclaredField("byKey")
            keyField.isAccessible = true
            val byKey = keyField[null] as HashMap<*, *>
            val nameField = Enchantment::class.java.getDeclaredField("byName")
            nameField.isAccessible = true
            val byName = nameField[null] as HashMap<*, *>
            for (enchantment in EEnchants.values()) {
                if (enchantment.key.startsWith(NamespacedKey.minecraft("bmc_").toString())) {
                    byKey.remove(enchantment.enchant?.key)
                    byName.remove(enchantment.enchant?.name)
                }
            }
        } catch (ignored: Exception) {
        }
    }

    fun hasEnchantment(item: ItemStack, enchantment: Enchantment?): Boolean {
        return item.enchantments.containsKey(enchantment)
    }

    fun addEnchantment(item: ItemStack, enchantment: Enchantment?, level: Int) {
        item.addUnsafeEnchantment(enchantment!!, level)
        item.updateCustomLore()
    }

    fun removeEnchantment(item: ItemStack, enchantment: Enchantment?) {
        item.removeEnchantment(enchantment!!)
        item.updateCustomLore()
    }

    fun getEnchantmentLevel(item: ItemStack, enchantment: Enchantment?): Int? {
        return item.enchantments[enchantment]
    }

    fun isCustom(enchantment: Enchantment): Boolean {
        return enchantment.key.toString().startsWith(NamespacedKey.minecraft("bmc_").toString())
    }
}